src/main/resources/admin/widgets/seo/seo.ts

```diff
@@ -50,128 +57,163 @@ export const get = (req: Request) => {
-			});
-			if (description === '') description = null;
-
-			const frontpageUrl = pageUrl({ path: siteOrNull?._path, type: "absolute" });
-			const absoluteUrl = pageUrl({ path: content._path, type: "absolute" });
-
-			let ogUrl: string;
-			if (appOrSiteConfig.baseUrl) {
-				ogUrl = prependBaseUrl({
-					baseUrl: appOrSiteConfig.baseUrl,
-					contentPath: content._path,
-					sitePath: siteOrNull?._path || ''
-				});
-			} else {
-				const justThePath = absoluteUrl.replace(frontpageUrl,'');
-				ogUrl = `[SITE_URL]${justThePath}`;
-			}
+	if (
+		startsWith(content.type,'media:')
+		|| arrayIncludes([
+			'portal:fragment',
+			'portal:template',
+			'portal:template-folder'
+		],(content.type))
+	) {
+		return {
+			body: `<widget class='error'>No metafields for contentType:${content.type}</widget>`,
+			contentType: CONTENT_TYPE,
+		};
+	}
 
-			let canonical = null;
-			const contentForCanonicalUrl = getContentForCanonicalUrl(content);
-			if (contentForCanonicalUrl) {
-				if (appOrSiteConfig.baseUrl) {
-					canonical = prependBaseUrl({
-						baseUrl: appOrSiteConfig.baseUrl,
-						contentPath: contentForCanonicalUrl
-							? contentForCanonicalUrl._path
-							: content._path,
-						sitePath: siteOrNull?._path || ''
-					});
-				} else {
-					const canonicalUrl = contentForCanonicalUrl
-						? pageUrl({ path: contentForCanonicalUrl._path, type: "absolute" })
-						: absoluteUrl;
-					const canonicalJustThePath = canonicalUrl.replace(frontpageUrl,'');
-					canonical = `[SITE_URL]${canonicalJustThePath}`;
-				}
-			}
+	const site: Site<MetafieldsSiteConfig>|null = getSiteByKey<MetafieldsSiteConfig>({ key: content._path });
+	DEBUG && log.debug('seo widget site:%s', toStr(site));
 
-			const imageUrl: string|null = getImageUrl({
-				appOrSiteConfig,
-				defaultImg: appOrSiteConfig.seoImage,
-				defaultImgPrescaled: appOrSiteConfig.seoImageIsPrescaled,
-				content,
-				siteOrNull,
+	if (!site) {
+		return {
+			body: `<widget class='error'>No metafields outside a site</widget>`,
+			contentType: CONTENT_TYPE,
+		};
+	}
+
+	const siteConfig = getSiteConfigOrNullFromContentKey(content._path)
+	DEBUG && log.debug('seo widget siteConfig:%s', toStr(siteConfig));
+	if (!siteConfig) {
+		return {
+			body: render( resolve('seo.html'), {}),
+			contentType: CONTENT_TYPE
+		};
+	}
+
+	const mergedConfig = getMergedConfig({siteConfig});
+	DEBUG && log.debug('seo widget mergedConfig:%s', toStr(mergedConfig));
+
+	if (!mergedConfig) {
+		return {
+			body: render( resolve('seo.html'), {}),
+			contentType: CONTENT_TYPE
+		};
+	}
+
+	const isFrontpage = site._path === content._path;
+	DEBUG && log.debug('seo widget isFrontpage:%s', isFrontpage);
+
+	const pageTitle = getPageTitle({
+		mergedConfig,
+		content,
+	});
+	const titleAppendix = getAppendix({
+		mergedConfig,
+		isFrontpage,
+		site,
+	});
+	let description = getMetaDescription({
+		mergedConfig,
+		content,
+		site,
+	});
+	if (description === '') description = null;
```

Comment:
This code is very similar to what we have here https://github.com/enonic/app-metafields/pull/146/files#diff-ed33235a54a6ff47c49c6d7261a38762b193d879cbdf06e7829607e48f55366fR87-R123. Not that it's a big deal but, considering complexity, maybe it's a good idea to refactor and use it once?

Reply:
Similar, but not the same. One runs in portal "context", the other does not. Also they return different stuff.